dbcache performance enhancements

ptschip opened the pull request:
This is currently fully functional, I just need to update some of the unit tests and should have this finished in the morning sometime.

I'm seeing very good performance even with a small 100MB dbcache.  Previously the dbcache was just getting cleared entirely every time the max limit was exceeded causing the next few blocks to be quite slow at validating while the cache was rebuilding.  All we do here is, when flushing the chainstate, keep cache entries that are not marked DIRTY and we also trim non-dirty entries if we exceed the max.

If you want to test with it use the following to see the log entries before and after flushing the chainstate.

debug=coindb
debug=thin

ptschip committed:
Performance improvement for in memory dbcache

When the cache is periodically flushed we don't want to
just clear the entire cache because there are still many
good entries there which then have to be pulled from disk
again if needed.  Rather, we wan't to just prune the entries
that are DIRTY which need to be flushed to disk and keep any
unmodified coins which can be used later during block
validation.

Add logprint for number of Trimmed entries

ptschip committed:
Continuously trim the dbcache while preventing continuous flushing

We want to make sure the cache is trimmed to it's max configured size
however we only want to do a full flush when needed which is usually
after a block is processed.  After a block is processed we check to
see if we are close to the max configured cache size and then proceed
to flush only when we're near.

ptschip committed:
Do not UnCache entries if they are already known.

There is no need to do this and instead we let the
pcoinsTip->Trim() take care of an excess entries in
the dbcache.

ptschip committed:
Change when we flush the chainstate to disk

Always flush the chainstate after completely processing a block
and updating the performance metrics - if - the chain is nearly
caught up.

Only flush as needed during IBD - if the chain is NOT nearly
caught up.

Doing the above keeps the chainstate always flushed and up to date
while at the same time not interfering with the performance\
metrics.

ptschip committed:
Increment Iterator and move locks

theslagter commented:
ptschip why you keep working on bitcoin unlimited? how many hours you spend on this per day?

ptschip committed:
Fix unit tests

ptschip commented:
taking out of WIP, all tests are passing now.

ptschip committed:
Periodically check and correct any drift in cachedCoinsUsage

While this should never happen, by checking periodically and correcting
drift the node can continue to operate and we only write an error to
the log.  Currently the check only happens once per day.

sickpig commented:
@ptschip need to be rebased cause it miss all the commit made from since Jun 16, namely:

```
e209fb8 gandrewstone    Merge pull request #675 from kyuupichan/CNetAddrService
046e241 gandrewstone    Merge pull request #674 from Justaphf/devWinDebug
be7c810 gandrewstone    Merge pull request #683 from kyuupichan/PV-fix
9fcf4b6 gandrewstone    Merge pull request #685 from ptschip/dev_headers
ba34c28 Peter Tschipper Remove DOS from unconnecting thinblock
4abd3b2 Peter Tschipper Do not continue if the header will not connect
fa9ab32 Neil Booth      PV: shut down threads only if initialized
79c23a0 Neil Booth      Move CNetAddr and CService to netaddress.{cpp,h}
b1d866b Justaphf        Fix debug build Pair ambiguity on Windows
```

sickpig commented:
@ptschip thanks

ptschip committed:
Make sure to flush the chainstate if reindexing or importing

If needed we have to ensure we flush the chainstate for these
two conditions until the chain is almost syncd.

ptschip commented:
@sickpig i added one small commit...it won't affect your general testing; it only impacts reindexing.

ptschip committed:
Use the more accurate DynamicMemoryUsage() when trimming the dbcache

Previously we were using the inner coins cache usage however this
does significantly lower the total bytes used which then causes
us to over trim the cache and with a subsequent lower performance.

ptschip commented:
rebased

gandrewstone commented on src/main.cpp:
```diff
@@ -1221,15 +1221,6 @@ bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
-                    vHashTxnToUncache.push_back(hash);
-                return state.Invalid(false, REJECT_ALREADY_KNOWN, "txn-already-known");
-            }
-
```
why is this deleted?

ptschip replied:
I took this out because it makes no sense...why would we delete a coin from the dbcache just because we already have it there...now we have it, now we don't...furthermore, why would we reject a txn based on what the state of the dbacahe is...the coin could be there already from finishing up IBD and there are ouputs left over in the dbcache or perhaps in the future for some other reason...if we're not going to allow a txn in the mempool it has to be because it's either an invalid txn or because we already have it in the mempool...checking the dbcache is not a good or reliable test IMO for whether the txn is in the mempool.

gandrewstone replied:
I think that this is an optimization.  it basically says "if the tx is already in the utxo cache, then clearly we already know about it".  IDK why you'd subsequently want to uncache the tx though... anyway I am fine with leaving it removed

gandrewstone commented on src/main.cpp:
```diff
@@ -3219,16 +3207,19 @@ bool static ConnectTip(CValidationState &state,
+    // Write the chain state to disk, if necessary, and only during IBD, reindex, or importing.
+    if (!IsChainNearlySyncd() || fReindex || fImporting)
+        if (!FlushStateToDisk(state, FLUSH_STATE_IF_NEEDED))
+            return false;
```
why is this not necessary during normal operation?

ptschip replied:
There are two reasons:  1) We flush ALWAYS now after every new block is processed (when processnewblock returns, at the end of HandleBlockMessageThread) during normal operation so checking again is redundant, furthermore, it injects performance timing issues if we flush at that point during ConnectTip  it can add a significant amount of time to the processing of a block in our statistics, when acutally, it has nothing to do with block processsing and should be excluded from the timing.
